src/machine.cpp

```diff
@@ -6157,6 +6201,68 @@ GcCallSite* resolveDynamic(Thread* t, GcInvocation* invocation)
     ++i;
   }
 
+  object varArgsArray_ = makeObjectArray(t, numVarArgs);
+  PROTECT(t, varArgsArray_);
+
+  if (numVarArgs > 0) {
+    assertT(t, i < bootstrapArray->length());
+    GcArray *varArgsArray = reinterpret_cast<GcArray*>(varArgsArray_);
```

Comment:
If avian's GC were a non-compacting GC, all it would need to know about is the set of live objects so it can figure out what's not live and free that.  If that were the case, you would be right.

However, avian's GC is actually a compacting GC (at least in the first generation), meaning that it not only needs to know about all the live objects, but also all the pointers thereto so that it can update them to the new location.

Now, it's possible that the compiler would actually combine those two `GcArray*`/`GcObject*` variables into one (since it knows they're the same anyway) - but I think that's very unlikely since it changes the semantics of the program if the pointer-to-pointer in the PROTECT were ever actually stored to somewhere in that function.  Without that optimization, only one variable would change, but with it, both would change. (Assuming we leave it as is.)

If there's a garbage collection while executing that function, the object pointed to by `varArgsArray` might be moved - and if `varArgsArray` is not PROTECTed, it won't be updated to the new location, leaving us with a pointer to who-knows-where.

------

Also, I just noticed you're doing a `reinterpret_cast` here, which works fine in the happy path.  If you were casting an unknown object rather than an object you created, I would be more particular about using the checked (at least in debug builds) `cast` function rather than the `reinterpret_cast` function.

I think the right solution is to `cast` / `reinterpret_cast` the original return value from `makeObjectArray`, eliminating the need for the second variable.  I would point out, however, that all the other cases where we're calling `makeObjectArray` and then initializing it, we're actually just using the unsafe/untyped `setField` function to do that, without any casts.  I don't remember exactly why I left those that way I left those that way, but it *might* be significant.

That said, if this works in practice and we have tests covering it (which it looks like we might not...), I'm standing on thin ice.

Would it be difficult to add a test or two that hits this code path?  Or should the existing tests hit this?